File: allcode/1900-1999/1970latestDayToCross.py
# 给你一个下标从 1 开始的二进制矩阵，其中 0 表示陆地，1 表示水域。同时给你 row 和 col 分别表示矩阵中行和列的数目。
#
# 一开始在第 0 天，整个 矩阵都是 陆地 。但每一天都会有一块新陆地被 水 淹没变成水域。给你一个下标从 1 开始的二维数组 cells ，其中 cells[i] = [ri, ci] 表示在第 i 天，第 ri 行 ci 列（下标都是从 1 开始）的陆地会变成 水域 （也就是 0 变成 1 ）。
#
# 你想知道从矩阵最 上面 一行走到最 下面 一行，且只经过陆地格子的 最后一天 是哪一天。你可以从最上面一行的 任意 格子出发，到达最下面一行的 任意 格子。你只能沿着 四个 基本方向移动（也就是上下左右）。
#
# 请返回只经过陆地格子能从最 上面 一行走到最 下面 一行的 最后一天 。
#
#
#
# 示例 1：
#
#
# 输入：row = 2, col = 2, cells = [[1,1],[2,1],[1,2],[2,2]]
# 输出：2
# 解释：上图描述了矩阵从第 0 天开始是如何变化的。
# 可以从最上面一行到最下面一行的最后一天是第 2 天。
# 示例 2：
#
#
# 输入：row = 2, col = 2, cells = [[1,1],[1,2],[2,1],[2,2]]
# 输出：1
# 解释：上图描述了矩阵从第 0 天开始是如何变化的。
# 可以从最上面一行到最下面一行的最后一天是第 1 天。
# 示例 3：
#
#
# 输入：row = 3, col = 3, cells = [[1,2],[2,1],[3,3],[2,2],[1,1],[1,3],[2,3],[3,2],[3,1]]
# 输出：3
# 解释：上图描述了矩阵从第 0 天开始是如何变化的。
# 可以从最上面一行到最下面一行的最后一天是第 3 天。
#
#
# 提示：
#
# 2 <= row, col <= 2 * 104
# 4 <= row * col <= 2 * 104
# cells.length == row * col
# 1 <= ri <= row
# 1 <= ci <= col
# cells 中的所有格子坐标都是 唯一 的。

from leetcode.allcode.competition.mypackage import *

class Solution:
    def latestDayToCross(self, row: int, col: int, cells: List[List[int]]) -> int:
        dir = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        n = row * col
        fa = list(range(n))
        # fa = {x: x for x in nums}  # 另一种写法，x不连续
        def find(x):
            if x != fa[x]:
                fa[x] = find(fa[x])
            return fa[x]
        def union(x, y):
            fa[find(y)] = find(x)

        for i in range(col):
            # 第一行和最后一行都在一个并查集中
            union(i, 0)
            union(col * (row - 1) + i, col * row - 1)

        vis = set()

        for i in range(row * col - 1, -1, -1):
            x, y = cells[i]
            x -= 1
            y -= 1
            # 倒序遍历
            for x0, y0 in dir:
                u, v = x + x0, y + y0
                if 0 <= u < row and 0 <= v < col and (u, v) in vis:
                    union(u * col + v, x * col + y)
            vis.add((x, y))
            if find(0) == find(col * row - 1):
                return i

so = Solution()
print(so.latestDayToCross(row = 2, col = 2, cells = [[1,1],[2,1],[1,2],[2,2]]))
print(so.latestDayToCross(row = 2, col = 2, cells = [[1,1],[1,2],[2,1],[2,2]]))
print(so.latestDayToCross(row = 3, col = 3, cells = [[1,2],[2,1],[3,3],[2,2],[1,1],[1,3],[2,3],[3,2],[3,1]]))

File: c++/src/1000-1999/1970latestDayToCross.cpp
#include "lc_pub.h"

class UnionFind {
    vector<int> fa; // 代表元
    vector<int> sz; // 集合大小

public:
    int cc; // 连通块个数

    UnionFind(int n) : fa(n), sz(n, 1), cc(n) {
        // 一开始有 n 个集合 {0}, {1}, ..., {n-1}
        // 集合 i 的代表元是自己，大小为 1
        ranges::iota(fa, 0); // iota(fa.begin(), fa.end(), 0);
    }

    // 返回 x 所在集合的代表元
    // 同时做路径压缩，也就是把 x 所在集合中的所有元素的 fa 都改成代表元
    int find(int x) {
        // 如果 fa[x] == x，则表示 x 是代表元
        if (fa[x] != x) {
            fa[x] = find(fa[x]); // fa 改成代表元
        }
        return fa[x];
    }

    // 判断 x 和 y 是否在同一个集合
    bool is_same(int x, int y) {
        // 如果 x 的代表元和 y 的代表元相同，那么 x 和 y 就在同一个集合
        // 这就是代表元的作用：用来快速判断两个元素是否在同一个集合
        return find(x) == find(y);
    }

    // 把 from 所在集合合并到 to 所在集合中
    // 返回是否合并成功
    bool merge(int from, int to) {
        int x = find(from), y = find(to);
        if (x == y) { // from 和 to 在同一个集合，不做合并
            return false;
        }
        fa[x] = y; // 合并集合。修改后就可以认为 from 和 to 在同一个集合了
        sz[y] += sz[x]; // 更新集合大小（注意集合大小保存在代表元上）
        // 无需更新 sz[x]，因为我们不用 sz[x] 而是用 sz[find(x)] 获取集合大小，但 find(x) == y，我们不会再访问 sz[x]
        cc--; // 成功合并，连通块个数减一
        return true;
    }

    // 返回 x 所在集合的大小
    int get_size(int x) {
        return sz[find(x)]; // 集合大小保存在代表元上
    }
};

class Solution {
public:
    int latestDayToCross(int row, int col, vector<vector<int>>& cells) {
        vector<pair<int, int>> dir={{1,0},{-1,0},{0,1},{0,-1}};
        int n = row*col;
        vector<vector<int>> grid(row, vector<int>(col, 1));
        UnionFind uf(n);
        for (int j=1;j<col;j++) {
            uf.merge(j-1,j);
            uf.merge((row-1)*col+j-1,(row-1)*col+j);
        }
        int m = cells.size();
        for (int i=m - 1;i>=0;i--) {
            int x=cells[i][0]-1,y=cells[i][1]-1;
            int z1=x*col+y;
            grid[x][y]=0;
            for (auto [dx,dy]: dir) {
                int u=x+dx,v=y+dy;
                if (0<=u&&u<row&&0<=v&&v<col&&grid[u][v]==0) {
                    int z2=u*col+v;
                    uf.merge(z1,z2);
                }
            }
            if (uf.is_same(0, n-1)) return i;
        }
        return 0;
    }
};

int main()
{
    cout<<"test let us start! %s" << __cplusplus <<std::endl;
    vector<int> nums={1,3};
    // vector<vector<int>> grid = parseGrid("[[1,1,1,-1,-1],[1,1,1,-1,-1],[-1,-1,-1,1,1],[1,1,1,1,-1],[-1,-1,-1,-1,-1]]");
    // cout << grid.size() << "  " << grid[0].size()<< endl;

    Solution so;
    return 0;
}
